doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+# Decision record: Learning To Rank
+
+**Date:** 2019-11-07
+
+The search team have decided to implement [Learning To Rank][].
+
+1. What is Learning To Rank?
+1. Motivation
+1. Implementation
+1. What success will look like
+
+[Learning To Rank]: https://en.wikipedia.org/wiki/Learning_to_rank
+
+## What is Learning To Rank?
+
+Learning To Rank (LTR) is the application of machine learning to create a
+ranking model that can be used to improve search relevance.
+
+This is a common approach for advanced search applications.
+
+The ranking of search results on websites is frequently determined by
+machine learning models. For example, Airbnb, Yelp, Wikipedia, Bloomberg,
+and commercial search engines.
+
+There is a good explanation on what LTR is here:
+
+- https://elasticsearch-learning-to-rank.readthedocs.io/en/latest/core-concepts.html
+- https://opensourceconnections.com/blog/2017/02/24/what-is-learning-to-rank/
+
+### How does Learning To Rank work?
+
+It's easiest to explain with an example:
+
+Say you searched for 'Harry Potter'. Search API might return the following results:
+
+1. HMRC Sign in
+2. Harry Potter World
+3. Pottery barn
+4. Harry Potter
+
+These are not good results.
+
+To assess search quality, we can assign relevancy judgements to the
+results of a query in the range 0-3, with 0 being irrelevant and 3
+being perfect. For example, for these results:
+
+```
+query         rating    document
+harry potter  0         HMRC Sign in
+harry potter  2         Harry Potter World
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+```
+
+We can then use a ranking quality metric, such as [normalised
+discounted cumulative gain (nDCG)](https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Normalized_DCG) to assign a score to
+this list of results.
+
+We can then use nDCG to tell us how good the results are.
+We might get a score of 0.3 for this result set (from 0 to 1).
+
+We could then investigate why the results are bad. Perhaps there is a
+best bet. Or popularity might be pushing HMRC docs up. We can fix these
+things, then check the new nDCG score, and we might get 0.7 for this query.
+
+In contrast, with Learning To Rank, search result ranks would be
+determined by a machine learning model without manually changing
+the boost of a field.
+
+With Learning To Rank we would go through the same process of
+rating the search results.
+
+```
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+...
+```
+
+However, after doing this, we would train a model with this data
+to rank the results better. The model optimises for nDCG.
+
+We train the model using using a combination of relevancy judgements
+and features: numeric measures which indicate how well we think the
+result matches the query.
+
+The text similarity scores Elasticsearch assigns to fields in the
+document are examples of features. The model tries to find correlations
```

Comment:
I think it would be helpful to have a section that lists the features you are considering using, and exploring which ones are likely to have the biggest influences on the model.

Reply:
I think listing the features we're using in documentation would be a good thing. At the moment I don't want to add much more on the implementation to the ADR, since it will get out of date quite soon. 

The features we've been experimenting with are: popularity, format, public_timestamp, es_score, title_score, description_score, indexable_content_score, all_searchable_text_score, title_length, description_length, link_length.